File: app/lib/models/chroma_fix_parameters.dart
import 'package:json_annotation/json_annotation.dart';

part 'chroma_fix_parameters.g.dart';

/// Chroma fix preset options.
enum ChromaFixPreset {
  @JsonValue('off')
  off,
  @JsonValue('vhsCleanup')
  vhsCleanup,
  @JsonValue('broadcastFix')
  broadcastFix,
  @JsonValue('analogRepair')
  analogRepair,
  @JsonValue('custom')
  custom,
}

/// Parameters for the chroma fix pass.
/// Includes FixChromaBleedingMod, LUTDeCrawl, and Vinverse filters.
@JsonSerializable()
class ChromaFixParameters {
  /// Whether this pass is enabled.
  final bool enabled;

  /// Preset level for simple mode.
  final ChromaFixPreset preset;

  // --- Chroma Shift (Y/C Delay) Parameters ---

  /// Whether to apply chroma shift (Y/C delay correction).
  final bool applyChromaShift;

  /// Horizontal chroma shift in luma pixels (negative = left, positive = right).
  final double chromaShiftH;

  /// Vertical chroma shift in luma pixels (negative = up, positive = down).
  final double chromaShiftV;

  // --- FixChromaBleedingMod Parameters ---

  /// Whether to apply chroma bleeding fix.
  final bool applyChromaBleedingFix;

  /// Chroma X offset correction.
  final int chromaBleedCx;

  /// Chroma Y offset correction.
  final int chromaBleedCy;

  /// Chroma blur strength (0.0 to 1.5+).
  final double chromaBleedCBlur;

  /// Fix strength (0.0 to 1.0).
  final double chromaBleedStrength;

  // --- LUTDeCrawl Parameters ---

  /// Whether to apply de-crawl (chroma crawl/dot crawl fix).
  final bool applyDeCrawl;

  /// Luma threshold for de-crawl.
  final int deCrawlYThresh;

  /// Chroma threshold for de-crawl.
  final int deCrawlCThresh;

  /// Maximum difference allowed.
  final int deCrawlMaxDiff;

  // --- LUTDeRainbow Parameters ---

  /// Whether to apply LUTDeRainbow (cross-luminance / rainbowing removal).
  ///
  /// Same 8-10 bit limit as LUTDeCrawl — havsfunc rejects anything above 10-bit
  /// outright, so the worker runs the pass at 10-bit and restores the source
  /// format afterwards.
  final bool applyDeRainbow;

  /// DeDot — temporal dot crawl / rainbow removal on both planes.
  final bool applyAutoChroma;
  final int autoChromaMaxShift;
  final double autoChromaAccuracy;
  final int autoChromaReferenceFrame;

  final bool applyDedot;
  final int dedotLuma2d;
  final int dedotLumaT;
  final int dedotChromaT1;
  final int dedotChromaT2;

  /// Chroma difference threshold for detecting rainbowing.
  final int deRainbowCThresh;

  /// Luma difference threshold. Areas moving more than this are left alone.
  final int deRainbowYThresh;

  /// Use the luma difference in the decision as well as chroma.
  final bool deRainbowUseLuma;

  /// Require both chroma planes to agree before treating a pixel.
  final bool deRainbowLinkUv;

  // --- Bifrost (temporal rainbow removal) ---

  /// Apply Bifrost. Where LUTDeRainbow decides within a frame, this compares
  /// across frames, so it catches rainbowing that shimmers rather than sits
  /// still. 8-bit only — the worker converts down and restores.
  final bool applyBifrost;

  /// Luma difference above which a block is treated as motion and left alone.
  final double bifrostLumaThresh;

  /// How many neighbouring blocks must agree before a pixel is treated.
  final int bifrostVariation;

  /// Treat the source as interlaced, comparing fields rather than frames.
  final bool bifrostInterlaced;

  // --- Vinverse Parameters ---

  /// Whether to apply Vinverse (inverted telecine/chroma fix).
  final bool applyVinverse;

  /// Spatial strength for Vinverse.
  final double vinverseSstr;

  /// Amount parameter for Vinverse (0-255).
  final int vinverseAmnt;

  const ChromaFixParameters({
    this.enabled = false,
    this.preset = ChromaFixPreset.off,
    // Chroma Shift defaults
    this.applyChromaShift = false,
    this.chromaShiftH = 0.0,
    this.chromaShiftV = 0.0,
    // FixChromaBleedingMod defaults
    this.applyChromaBleedingFix = false,
    this.chromaBleedCx = 4,
    this.chromaBleedCy = 4,
    this.chromaBleedCBlur = 0.7,
    this.chromaBleedStrength = 0.8,
    // LUTDeCrawl defaults
    this.applyDeCrawl = false,
    this.deCrawlYThresh = 10,
    this.deCrawlCThresh = 10,
    this.deCrawlMaxDiff = 50,
    // Vinverse defaults
    this.applyDeRainbow = false,
    this.applyAutoChroma = false,
    this.autoChromaMaxShift = 2,
    this.autoChromaAccuracy = 0.25,
    this.autoChromaReferenceFrame = 0,
    this.applyDedot = false,
    this.dedotLuma2d = 20,
    this.dedotLumaT = 20,
    this.dedotChromaT1 = 15,
    this.dedotChromaT2 = 5,
    this.deRainbowCThresh = 10,
    this.deRainbowYThresh = 10,
    this.deRainbowUseLuma = true,
    this.deRainbowLinkUv = true,
    this.applyBifrost = false,
    this.bifrostLumaThresh = 10.0,
    this.bifrostVariation = 5,
    this.bifrostInterlaced = true,
    this.applyVinverse = false,
    this.vinverseSstr = 2.7,
    this.vinverseAmnt = 255,
  });

  /// Create parameters from a preset.
  factory ChromaFixParameters.fromPreset(ChromaFixPreset preset) {
    switch (preset) {
      case ChromaFixPreset.off:
        return const ChromaFixParameters(
          enabled: false,
          preset: ChromaFixPreset.off,
        );
      case ChromaFixPreset.vhsCleanup:
        return const ChromaFixParameters(
          enabled: true,
          preset: ChromaFixPreset.vhsCleanup,
          applyChromaBleedingFix: true,
          chromaBleedCBlur: 0.8,
          chromaBleedStrength: 0.8,
          applyVinverse: true,
          vinverseSstr: 2.7,
        );
      case ChromaFixPreset.broadcastFix:
        return const ChromaFixParameters(
          enabled: true,
          preset: ChromaFixPreset.broadcastFix,
          applyDeCrawl: true,
          deCrawlYThresh: 12,
          deCrawlCThresh: 12,
        );
      case ChromaFixPreset.analogRepair:
        return const ChromaFixParameters(
          enabled: true,
          preset: ChromaFixPreset.analogRepair,
          applyChromaBleedingFix: true,
          chromaBleedCBlur: 1.0,
          chromaBleedStrength: 1.0,
          applyDeCrawl: true,
          applyVinverse: true,
        );
      case ChromaFixPreset.custom:
        return const ChromaFixParameters(
          enabled: true,
          preset: ChromaFixPreset.custom,
        );
    }
  }

  ChromaFixParameters copyWith({
    bool? enabled,
    ChromaFixPreset? preset,
    bool? applyChromaShift,
    double? chromaShiftH,
    double? chromaShiftV,
    bool? applyChromaBleedingFix,
    int? chromaBleedCx,
    int? chromaBleedCy,
    double? chromaBleedCBlur,
    double? chromaBleedStrength,
    bool? applyDeCrawl,
    int? deCrawlYThresh,
    int? deCrawlCThresh,
    int? deCrawlMaxDiff,
    bool? applyDeRainbow,
    int? deRainbowCThresh,
    int? deRainbowYThresh,
    bool? deRainbowUseLuma,
    bool? deRainbowLinkUv,
    bool? applyBifrost,
    double? bifrostLumaThresh,
    int? bifrostVariation,
    bool? bifrostInterlaced,
    bool? applyVinverse,
    double? vinverseSstr,
    int? vinverseAmnt,
  }) {
    return ChromaFixParameters(
      enabled: enabled ?? this.enabled,
      preset: preset ?? this.preset,
      applyChromaShift: applyChromaShift ?? this.applyChromaShift,
      chromaShiftH: chromaShiftH ?? this.chromaShiftH,
      chromaShiftV: chromaShiftV ?? this.chromaShiftV,
      applyChromaBleedingFix: applyChromaBleedingFix ?? this.applyChromaBleedingFix,
      chromaBleedCx: chromaBleedCx ?? this.chromaBleedCx,
      chromaBleedCy: chromaBleedCy ?? this.chromaBleedCy,
      chromaBleedCBlur: chromaBleedCBlur ?? this.chromaBleedCBlur,
      chromaBleedStrength: chromaBleedStrength ?? this.chromaBleedStrength,
      applyDeCrawl: applyDeCrawl ?? this.applyDeCrawl,
      deCrawlYThresh: deCrawlYThresh ?? this.deCrawlYThresh,
      deCrawlCThresh: deCrawlCThresh ?? this.deCrawlCThresh,
      deCrawlMaxDiff: deCrawlMaxDiff ?? this.deCrawlMaxDiff,
      applyDeRainbow: applyDeRainbow ?? this.applyDeRainbow,
      deRainbowCThresh: deRainbowCThresh ?? this.deRainbowCThresh,
      deRainbowYThresh: deRainbowYThresh ?? this.deRainbowYThresh,
      deRainbowUseLuma: deRainbowUseLuma ?? this.deRainbowUseLuma,
      deRainbowLinkUv: deRainbowLinkUv ?? this.deRainbowLinkUv,
      applyBifrost: applyBifrost ?? this.applyBifrost,
      bifrostLumaThresh: bifrostLumaThresh ?? this.bifrostLumaThresh,
      bifrostVariation: bifrostVariation ?? this.bifrostVariation,
      bifrostInterlaced: bifrostInterlaced ?? this.bifrostInterlaced,
      applyVinverse: applyVinverse ?? this.applyVinverse,
      vinverseSstr: vinverseSstr ?? this.vinverseSstr,
      vinverseAmnt: vinverseAmnt ?? this.vinverseAmnt,
    );
  }

  /// Get a human-readable summary of the current settings.
  String get summary {
    if (!enabled) return 'Off';
    if (preset != ChromaFixPreset.custom && preset != ChromaFixPreset.off) {
      switch (preset) {
        case ChromaFixPreset.vhsCleanup:
          return 'VHS Cleanup';
        case ChromaFixPreset.broadcastFix:
          return 'Broadcast Fix';
        case ChromaFixPreset.analogRepair:
          return 'Analog Repair';
        default:
          return preset.name;
      }
    }
    final fixes = <String>[];
    if (applyChromaShift) fixes.add('Shift');
    if (applyChromaBleedingFix) fixes.add('Bleed');
    if (applyDeCrawl) fixes.add('Crawl');
    if (applyVinverse) fixes.add('Vinv');
    return fixes.isEmpty ? 'Custom' : fixes.join('+');
  }

  factory ChromaFixParameters.fromJson(Map<String, dynamic> json) =>
      _$ChromaFixParametersFromJson(json);

  Map<String, dynamic> toJson() => _$ChromaFixParametersToJson(this);
}

File: worker/src/models/chroma_fix_parameters.rs
//! Chroma fix parameters for video processing.

use serde::{Deserialize, Serialize};

/// Chroma fix preset options.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub enum ChromaFixPreset {
    #[default]
    Off,
    VhsCleanup,
    BroadcastFix,
    AnalogRepair,
    Custom,
}

/// Parameters for the chroma fix pass.
/// Includes FixChromaBleedingMod, LUTDeCrawl, and Vinverse filters.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChromaFixParameters {
    /// Whether this pass is enabled.
    #[serde(default)]
    pub enabled: bool,

    /// Preset level for simple mode.
    #[serde(default)]
    pub preset: ChromaFixPreset,

    // --- Chroma Shift (Y/C Delay) Parameters ---

    /// Whether to apply chroma shift (Y/C delay correction).
    #[serde(default)]
    pub apply_chroma_shift: bool,

    /// Horizontal chroma shift in luma pixels (negative = left, positive = right).
    #[serde(default)]
    pub chroma_shift_h: f64,

    /// Vertical chroma shift in luma pixels (negative = up, positive = down).
    #[serde(default)]
    pub chroma_shift_v: f64,

    // --- FixChromaBleedingMod Parameters ---

    /// Whether to apply chroma bleeding fix.
    #[serde(default)]
    pub apply_chroma_bleeding_fix: bool,

    /// Chroma X offset correction.
    #[serde(default = "default_chroma_bleed_offset")]
    pub chroma_bleed_cx: i32,

    /// Chroma Y offset correction.
    #[serde(default = "default_chroma_bleed_offset")]
    pub chroma_bleed_cy: i32,

    /// Chroma blur strength (0.0 to 1.5+).
    #[serde(default = "default_chroma_bleed_blur")]
    pub chroma_bleed_c_blur: f64,

    /// Fix strength (0.0 to 1.0).
    #[serde(default = "default_chroma_bleed_strength")]
    pub chroma_bleed_strength: f64,

    // --- LUTDeCrawl Parameters ---

    /// Whether to apply de-crawl (chroma crawl/dot crawl fix).
    #[serde(default)]
    pub apply_de_crawl: bool,

    /// Luma threshold for de-crawl.
    #[serde(default = "default_de_crawl_thresh")]
    pub de_crawl_y_thresh: i32,

    /// Chroma threshold for de-crawl.
    #[serde(default = "default_de_crawl_thresh")]
    pub de_crawl_c_thresh: i32,

    /// Maximum difference allowed.
    #[serde(default = "default_de_crawl_max_diff")]
    pub de_crawl_max_diff: i32,

    // --- LUTDeRainbow (cross-luminance / rainbowing) ---

    /// Apply LUTDeRainbow.
    #[serde(default)]
    pub apply_de_rainbow: bool,

    /// DeDot — temporal dot crawl / rainbow removal on both planes.
    #[serde(default)]
    pub apply_dedot: bool,

    /// Measure the chroma misalignment and correct it automatically.
    #[serde(default)]
    pub apply_auto_chroma: bool,
    /// Largest shift to search for, in pixels.
    #[serde(default = "default_acf_max_shift")]
    pub auto_chroma_max_shift: i32,
    /// Sub-pixel search step.
    #[serde(default = "default_acf_accuracy")]
    pub auto_chroma_accuracy: f64,
    /// Measure once on this frame (-1 measures every frame, ~23x the cost).
    #[serde(default)]
    pub auto_chroma_reference_frame: i32,
    /// Spatial luma threshold (0-510).
    #[serde(default = "default_dedot_luma_2d")]
    pub dedot_luma_2d: i32,
    /// Temporal luma threshold (0-255).
    #[serde(default = "default_dedot_luma_t")]
    pub dedot_luma_t: i32,
    /// Chroma threshold 1 (0-255).
    #[serde(default = "default_dedot_chroma_t1")]
    pub dedot_chroma_t1: i32,
    /// Chroma threshold 2 (0-255). 255 bypasses chroma entirely.
    #[serde(default = "default_dedot_chroma_t2")]
    pub dedot_chroma_t2: i32,

    /// Chroma difference threshold for detecting rainbowing.
    #[serde(default = "default_de_rainbow_cthresh")]
    pub de_rainbow_c_thresh: i32,

    /// Luma difference threshold. Areas moving more than this are left alone.
    #[serde(default = "default_de_rainbow_ythresh")]
    pub de_rainbow_y_thresh: i32,

    /// Use the luma difference in the decision as well as chroma.
    #[serde(default = "default_true")]
    pub de_rainbow_use_luma: bool,

    /// Require both chroma planes to agree before treating a pixel.
    #[serde(default = "default_true")]
    pub de_rainbow_link_uv: bool,

    // --- Bifrost (temporal rainbow removal) ---

    /// Apply Bifrost. Where LUTDeRainbow works within a frame, this compares
    /// across frames, so it catches rainbowing that shimmers rather than sits
    /// still.
    #[serde(default)]
    pub apply_bifrost: bool,

    /// Luma difference above which a block is treated as motion and left alone.
    #[serde(default = "default_bifrost_luma_thresh")]
    pub bifrost_luma_thresh: f64,

    /// How many neighbouring blocks must agree before a pixel is treated
    /// (0-3). Higher is more conservative.
    #[serde(default = "default_bifrost_variation")]
    pub bifrost_variation: i32,

    /// Treat the source as interlaced, comparing fields rather than frames.
    #[serde(default = "default_true")]
    pub bifrost_interlaced: bool,

    // --- Vinverse Parameters ---

    /// Whether to apply Vinverse (inverted telecine/chroma fix).
    #[serde(default)]
    pub apply_vinverse: bool,

    /// Spatial strength for Vinverse.
    #[serde(default = "default_vinverse_sstr")]
    pub vinverse_sstr: f64,

    /// Amount parameter for Vinverse (0-255).
    #[serde(default = "default_255")]
    pub vinverse_amnt: i32,
}

fn default_chroma_bleed_offset() -> i32 { 4 }
fn default_chroma_bleed_blur() -> f64 { 0.7 }
fn default_chroma_bleed_strength() -> f64 { 0.8 }
fn default_de_crawl_thresh() -> i32 { 10 }
fn default_de_crawl_max_diff() -> i32 { 50 }
fn default_true() -> bool { true }
fn default_de_rainbow_cthresh() -> i32 { 10 }
fn default_de_rainbow_ythresh() -> i32 { 10 }
fn default_bifrost_luma_thresh() -> f64 { 10.0 }
fn default_bifrost_variation() -> i32 { 5 }
fn default_vinverse_sstr() -> f64 { 2.7 }
fn default_255() -> i32 { 255 }

fn default_acf_max_shift() -> i32 { 2 }
fn default_acf_accuracy() -> f64 { 0.25 }
fn default_dedot_luma_2d() -> i32 { 20 }
fn default_dedot_luma_t() -> i32 { 20 }
fn default_dedot_chroma_t1() -> i32 { 15 }
fn default_dedot_chroma_t2() -> i32 { 5 }

impl Default for ChromaFixParameters {
    fn default() -> Self {
        Self {
            enabled: false,
            preset: ChromaFixPreset::default(),
            apply_chroma_shift: false,
            chroma_shift_h: 0.0,
            chroma_shift_v: 0.0,
            apply_chroma_bleeding_fix: false,
            chroma_bleed_cx: default_chroma_bleed_offset(),
            chroma_bleed_cy: default_chroma_bleed_offset(),
            chroma_bleed_c_blur: default_chroma_bleed_blur(),
            chroma_bleed_strength: default_chroma_bleed_strength(),
            apply_de_crawl: false,
            apply_de_rainbow: false,
            apply_dedot: false,
            apply_auto_chroma: false,
            auto_chroma_max_shift: default_acf_max_shift(),
            auto_chroma_accuracy: default_acf_accuracy(),
            auto_chroma_reference_frame: 0,
            dedot_luma_2d: default_dedot_luma_2d(),
            dedot_luma_t: default_dedot_luma_t(),
            dedot_chroma_t1: default_dedot_chroma_t1(),
            dedot_chroma_t2: default_dedot_chroma_t2(),
            de_rainbow_c_thresh: default_de_rainbow_cthresh(),
            de_rainbow_y_thresh: default_de_rainbow_ythresh(),
            de_rainbow_use_luma: true,
            de_rainbow_link_uv: true,
            apply_bifrost: false,
            bifrost_luma_thresh: default_bifrost_luma_thresh(),
            bifrost_variation: default_bifrost_variation(),
            bifrost_interlaced: true,
            de_crawl_y_thresh: default_de_crawl_thresh(),
            de_crawl_c_thresh: default_de_crawl_thresh(),
            de_crawl_max_diff: default_de_crawl_max_diff(),
            apply_vinverse: false,
            vinverse_sstr: default_vinverse_sstr(),
            vinverse_amnt: default_255(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_parameters() {
        let params = ChromaFixParameters::default();
        assert!(!params.enabled);
        assert_eq!(params.preset, ChromaFixPreset::Off);
        assert!(!params.apply_chroma_shift);
        assert_eq!(params.chroma_shift_h, 0.0);
        assert_eq!(params.chroma_shift_v, 0.0);
        assert!(!params.apply_chroma_bleeding_fix);
        assert!(!params.apply_de_crawl);
        assert!(!params.apply_vinverse);
    }

    #[test]
    fn test_automatic_alignment_supersedes_the_manual_shift() {
        let mut params = ChromaFixParameters {
            apply_chroma_shift: true,
            chroma_shift_h: 2.0,
            ..ChromaFixParameters::default()
        };
        assert!(params.effective_apply_chroma_shift());

        params.apply_auto_chroma = true;
        assert!(
            !params.effective_apply_chroma_shift(),
            "the automatic pass already corrects the shift; applying the manual \
             one on top would double-correct it"
        );
        // The user's own setting survives, so unticking automatic restores it.
        assert!(params.apply_chroma_shift);
        assert_eq!(params.chroma_shift_h, 2.0);
    }

    #[test]
    fn test_serialization() {
        let params = ChromaFixParameters::default();
        let json = serde_json::to_string(&params).unwrap();
        assert!(json.contains("\"enabled\":false"));
        assert!(json.contains("\"chromaBleedCx\":4"));
    }
}

impl ChromaFixParameters {
    /// Bifrost's `variation`, clamped to the range it accepts.
    pub fn bifrost_effective_variation(&self) -> i32 {
        self.bifrost_variation.clamp(0, 10)
    }

    /// Whether the manual Y/C delay shift reaches the script.
    ///
    /// Automatic alignment measures the misalignment and corrects it, and its
    /// block runs *before* the manual shift — so with both set the picture is
    /// shifted twice, the second time by a number the user guessed on top of a
    /// correction that was already measured. The automatic measurement wins:
    /// the panel hides the manual controls while it is on
    /// (`chroma_fixes.json`, `visibleWhen: {applyAutoChroma: false}`) and this
    /// keeps the generated script agreeing with what the panel shows.
    ///
    /// It deliberately does not clear `apply_chroma_shift` in the job, so
    /// turning automatic alignment off restores what the user set by hand.
    /// `ChromaFixParameters.effectiveApplyChromaShift` is the Dart twin.
    pub fn effective_apply_chroma_shift(&self) -> bool {
        self.apply_chroma_shift && !self.apply_auto_chroma
    }
}
